### PR TITLE
Feature/orco 18

### PR DESCRIPTION
**Pull request by cpjulia**
### Scope & Purpose

OBS.: ignoring clang-formatting for the moment because of headers

This PR is related to https://arangodb.atlassian.net/browse/APM-24. Upon user demand, log messages can be displayed together with some other useful parameters, e.g., the name of the database that is being used at the moment, the username, the query id, and so on. 
There are some predefined parameters that we consider displaying, but, for the moment, only the database, username and url are being displayed. This is the basic usage of it. The other parameters are to be implemented. 
The usage upon starting the server is, for example: 
`arangod --log.structured-param database=true --log.structured-param username`
As shown in the example above, the user can either provide just the string with the parameter's name or also the boolean value for the parameter. If just the parameter's name is provided, it's considered that its value is true. 
There's another use case, which is when the server is running, and the user wants to change the parameter list dinamically. 
That would be, for example:
`arango.PUT("/_admin/log/structured", {"database": true, "url": false})`
here we're adding the database parameter and, if there was a url parameter set upon server start, it's removing it from the parameter list. 

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

No backports required.

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [x] Docs PR: to be added here 
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket number: https://arangodb.atlassian.net/browse/APM-24
- [ ] Design document: 

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

File: arangod/GeneralServer/RestHandler.h [hunk 1]
```diff
@@ -233,6 +236,11 @@ class RestHandler : public std::enable_shared_from_this<RestHandler> {
 
   RequestLane _lane;
 
+  std::shared_ptr<LogContext::Values> _logContextScopeValues;
+  LogContext::EntryPtr _logContextEntry;
+  static constexpr char UrlName[] = "url";
+  static constexpr char UserName[] = "username";
```
Comment (jsteemann) [inline arangod/GeneralServer/RestHandler.h:242]: Didn't we say all the parameters should be placed in a central header file?

File: arangod/RestHandler/RestAdminLogHandler.cpp [hunk 2]
```diff
@@ -113,13 +113,19 @@ RestStatus RestAdminLogHandler::execute() {
...
       generateError(rest::ResponseCode::BAD,
                     TRI_ERROR_HTTP_SUPERFLUOUS_SUFFICES,
                     "superfluous suffix, expecting /_admin/log/entries");
     }
-  } else {
-    handleLogLevel();
+  } else if (type == rest::RequestType::PUT) {
+    if (suffixes.size() == 1 && suffixes[0] == "level") {
+      handleLogLevel();
+    } else {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminLogHandler.cpp:126]: We should also check the number of suffixes here and the URL part "structured". Otherwise users could call PUT `/_admin/log/testi/1234` and it would work.

File: arangod/RestHandler/RestAdminLogHandler.cpp [hunk 3]
```diff
@@ -472,3 +478,63 @@ void RestAdminLogHandler::handleLogLevel() {
...
+
+  // was validated earlier
+  TRI_ASSERT(!suffixes.empty());
+
+  if (suffixes[0] != "structured") {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_SUPERFLUOUS_SUFFICES,
+                  "superfluous suffix, expecting /_admin/log/structured");
+    return;
+  }
+
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminLogHandler.cpp:493]: This shouldn't be done here, but in the caller.

File: arangod/RestHandler/RestLogHandler.cpp [hunk 4]
```diff
@@ -71,6 +73,41 @@ RestStatus RestLogHandler::executeByMethod(
   return RestStatus::DONE;
 }
 
+RestStatus RestLogHandler::handlePutRequest(
```
Comment (jsteemann) [inline arangod/RestHandler/RestLogHandler.cpp:76]: What is the rationale for this change? IMO this is a handler for replication2...

File: lib/Logger/LogStructuredParamsAllowList.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#pragma once
+
+#include <string>
+#include <unordered_set>
+
+namespace arangodb {
+namespace structuredParams {
+const std::unordered_set<std::string> allowList = {
+    "database", "username", "url",         "collection", "queryid",
+    "viewId",   "indexid",  "replication", "pregelid",   "transactionid"};
```
Comment (jsteemann) [inline lib/Logger/LogStructuredParamsAllowList.h:33]: capitalization is inconsistent: `viewId` vs. `indexid`. Please use consistent capitalization, i.e. `indexId`, `pregeld`, `transactionId`, `queryId`.

File: lib/Logger/Logger.cpp [hunk 6]
```diff
@@ -218,12 +227,80 @@ void Logger::setLogLevel(std::string const& levelName) {
...
 
+void Logger::setLogStructuredParam(
+    std::pair<std::string, bool> const& paramAndValue) {
+  auto const& paramName = paramAndValue.first;
+  bool value = paramAndValue.second;
+  if (value) {
+    if (auto it = _structuredLogParams.find(paramName);
+        it == _structuredLogParams.end()) {
+      _structuredLogParams.emplace(paramName);
+    }
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:238]:
```suggestion
      _structuredLogParams.emplace(paramName);
```

File: lib/Logger/Logger.cpp [hunk 7]
```diff
@@ -218,12 +227,80 @@ void Logger::setLogLevel(std::string const& levelName) {
...
+  } else {
+    if (auto it = _structuredLogParams.find(paramName);
+        it != _structuredLogParams.end()) {
+      _structuredLogParams.erase(it);
+    }
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:243]:
```suggestion
    _structuredLogParams.erase(paramName);
```

File: lib/Logger/Logger.cpp [hunk 8]
```diff
@@ -218,12 +227,80 @@ void Logger::setLogLevel(std::string const& levelName) {
...
+    std::string l = StringUtils::tolower(param);
+    std::vector<std::string> v = StringUtils::split(l, '=');
+    size_t vSize = v.size();
+    if (!vSize || vSize > 2) {
+      LOG_TOPIC("4d971", ERR, arangodb::Logger::FIXME)
+          << "strange log attribute and value set '" + param + "'";
+    } else {
+      StringUtils::trimInPlace(v[0]);
+      if (!structuredParams::allowList.contains(v[0])) {
+        LOG_TOPIC("c4c17", ERR, arangodb::Logger::FIXME)
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:266]:
```suggestion
        LOG_TOPIC("c4c17", WARN, arangodb::Logger::FIXME)
```

File: lib/Logger/Logger.cpp [hunk 9]
```diff
@@ -218,12 +227,80 @@ void Logger::setLogLevel(std::string const& levelName) {
...
+      if (vSize == 2) {
+        StringUtils::trimInPlace(v[1]);
+      }
+      if (vSize == 1 || v[1] == "true") {
+        validParams[v[0]] = true;
+      } else {
+        if (v[1] == "false") {
+          validParams[v[0]] = false;
+        } else {
+          LOG_TOPIC("5d210", ERR, arangodb::Logger::FIXME)
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:279]:
```suggestion
          LOG_TOPIC("5d210", WARN, arangodb::Logger::FIXME)
```

File: lib/Logger/Logger.cpp [hunk 10]
```diff
@@ -553,21 +630,25 @@ void Logger::log(char const* logid, char const* function, char const* file,
...
-          }
-        });
-    logContext.visit(visitor);
+    {
+      READ_LOCKER(guard, _structuredParamsLock);
+      // meta data from log context
+      LogContext::OverloadVisitor visitor([&out, &dumper](
+                                              std::string_view const& key,
+                                              auto&& value) {
+        out.push_back(',');
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:639]: This is missing the following check
```
        if (!_structuredLogParams.contains(key.data())) {
          return;
        }
```

File: lib/Logger/Logger.cpp [hunk 11]
```diff
@@ -695,21 +776,28 @@ void Logger::log(char const* logid, char const* function, char const* file,
...
-            out.append(value);
-          } else {
-            out.append(std::to_string(value));
-          }
-          out.append("] ", 2);
-        });
-    logContext.visit(visitor);
+    {
+      READ_LOCKER(guard, _structuredParamsLock);
+      // JULIA
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:781]:
```suggestion
```

File: lib/Logger/Logger.h [hunk 12]
```diff
@@ -255,11 +259,21 @@ class Logger {
...
 
+  static std::unordered_map<std::string, bool> const filterInvalidParams(
+      std::vector<std::string> const&);
+  static void setLogStructuredParams(std::vector<std::string> const&);
+  static void setLogStructuredParamsOnServerStart(
+      std::vector<std::string> const&);
+  static void setLogStructuredParam(std::pair<std::string, bool> const&);
+
+  //  static void setLogStructuredParam(std::string const&);
+
```
Comment (jsteemann) [inline lib/Logger/Logger.h:276]:
```suggestion
```

File: lib/Logger/LoggerFeature.cpp [hunk 13]
```diff
@@ -56,6 +56,7 @@
 #include "ProgramOptions/Option.h"
 #include "ProgramOptions/Parameters.h"
 #include "ProgramOptions/ProgramOptions.h"
+#include <velocypack/Parser.h>
```
Comment (jsteemann) [inline lib/Logger/LoggerFeature.cpp:59]:
```suggestion
```

File: lib/Logger/LoggerFeature.h [hunk 14]
```diff
@@ -29,6 +29,7 @@
 #include <vector>
 
 #include "ApplicationFeatures/ApplicationFeature.h"
+#include <velocypack/Builder.h>
```
Comment (jsteemann) [inline lib/Logger/LoggerFeature.h:32]: Why is this necessary here?

File: tests/js/server/aql/aql-queries-simple.js [hunk 15]
```diff
@@ -1355,6 +1355,9 @@ function ahuacatlQuerySimpleTestSuite () {
       }
       q += `RETURN v${cnt - 1})[0]\n`;
       q += "RETURN i + x";
+      require('console').log(q);
+      const db = require('@arangodb').db;
+      db._explain(q);
```
Comment (jsteemann) [inline tests/js/server/aql/aql-queries-simple.js:1360]: Unrelated change and does not belong in this PR.

File: arangod/RestHandler/RestAdminLogHandler.cpp [hunk 16]
```diff
@@ -472,3 +478,63 @@ void RestAdminLogHandler::handleLogLevel() {
...
+  } else if (type == rest::RequestType::PUT) {
+    // set log level
+    bool parseSuccess = false;
+    VPackSlice slice = this->parseVPackBody(parseSuccess);
+    if (!parseSuccess) {
+      return;  // error message generated in parseVPackBody
+    }
+
+    if (slice.isString()) {
+      Logger::setLogStructuredParams({slice.copyString()});
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminLogHandler.cpp:515]: Should the REST API actually be callable with a string input? I don't think this is required functionality.

File: arangod/RestHandler/RestAdminLogHandler.cpp [hunk 17]
```diff
@@ -472,3 +478,63 @@ void RestAdminLogHandler::handleLogLevel() {
...
+    } else if (slice.isObject()) {
+      for (auto it : VPackObjectIterator(slice)) {
+        if (it.value.isBoolean()) {
+          std::string const l = it.key.copyString() + "=" +
+                                (it.value.getBoolean() ? "true" : "false");
+          Logger::setLogStructuredParams({l});
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminLogHandler.cpp:521]: This looks a bit strange.

File: CHANGELOG [hunk 18]
```diff
@@ -10,6 +17,7 @@ devel
   the read-only mode will be enforced.
   Upgrade procedure for a deployment without license will not take upgrade
   period into account for the read-only mode will enforcement.
+>>>>>>> 2c185436e101003640f390c93fa138583220a563
```
Comment (jsteemann) [inline CHANGELOG:20]:
```suggestion
```

File: arangod/GeneralServer/RestHandler.h [hunk 19]
```diff
@@ -120,10 +121,12 @@ class RestHandler : public std::enable_shared_from_this<RestHandler> {
 
   RequestLane determineRequestLane();
 
-  virtual void prepareExecute(bool isContinue) {}
+  virtual void prepareExecute(bool isContinue);
+  // virtual void executeWrapper() = { Scope.... execute()}; ScopeValue
+  // newValues(..............); execute(); }
```
Comment (jsteemann) [inline arangod/GeneralServer/RestHandler.h:126]:
```suggestion
```

File: arangod/GeneralServer/RestHandler.h (hunk 1 above)
Comment (cpjulia) [reply arangod/GeneralServer/RestHandler.h]: Addressed in latest commit

File: arangod/RestHandler/RestAdminLogHandler.cpp (hunk 2 above)
Comment (cpjulia) [reply arangod/RestHandler/RestAdminLogHandler.cpp]: Addressed in latest commit

File: arangod/RestHandler/RestAdminLogHandler.cpp (hunk 3 above)
Comment (cpjulia) [reply arangod/RestHandler/RestAdminLogHandler.cpp]: Addressed

File: arangod/RestHandler/RestVocbaseBaseHandler.h [hunk 20]
```diff
@@ -221,6 +227,11 @@ class RestVocbaseBaseHandler : public RestBaseHandler {
 
   /// @brief the vocbase, managed by VocbaseContext
   TRI_vocbase_t& _vocbase;
+
+ private:
+  static constexpr char DatabaseName[] = "database";
```
[the remark this answers is not in the document]
Comment (cpjulia) [reply arangod/RestHandler/RestVocbaseBaseHandler.h]: Addressed

File: lib/Logger/LogStructuredParamsAllowList.h (hunk 5 above)
Comment (cpjulia) [reply lib/Logger/LogStructuredParamsAllowList.h]: Addressed

File: lib/Logger/Logger.cpp (hunk 6 above)
Comment (cpjulia) [reply lib/Logger/Logger.cpp]: Addressed

File: lib/Logger/Logger.cpp (hunk 7 above)
Comment (cpjulia) [reply lib/Logger/Logger.cpp]: Addressed

File: lib/Logger/Logger.cpp [hunk 21]
```diff
@@ -218,12 +227,80 @@ void Logger::setLogLevel(std::string const& levelName) {
...
 
+std::unordered_map<std::string, bool> const Logger::filterInvalidParams(
+    std::vector<std::string> const& params) {
+  std::unordered_map<std::string, bool> validParams;
+  for (auto const& param : params) {
+    std::string l = StringUtils::tolower(param);
+    std::vector<std::string> v = StringUtils::split(l, '=');
+    size_t vSize = v.size();
+    if (!vSize || vSize > 2) {
+      LOG_TOPIC("4d971", ERR, arangodb::Logger::FIXME)
```
[the remark this answers is not in the document]
Comment (cpjulia) [reply lib/Logger/Logger.cpp]: Addressed

File: lib/Logger/Logger.cpp (hunk 8 above)
Comment (cpjulia) [reply lib/Logger/Logger.cpp]: Addressed

File: lib/Logger/Logger.cpp (hunk 10 above)
Comment (cpjulia) [reply lib/Logger/Logger.cpp:636]: Addressed

File: lib/Logger/Logger.cpp (hunk 11 above)
Comment (cpjulia) [reply lib/Logger/Logger.cpp]: Addressed

File: lib/Logger/Logger.h (hunk 12 above)
Comment (cpjulia) [reply lib/Logger/Logger.h]: Addressed

File: lib/Logger/LoggerFeature.cpp (hunk 13 above)
Comment (cpjulia) [reply lib/Logger/LoggerFeature.cpp]: Addressed

File: arangod/RestHandler/RestAdminLogHandler.cpp (hunk 17 above)
Comment (cpjulia) [reply arangod/RestHandler/RestAdminLogHandler.cpp:516]: This is because I need to know whether the parameters supplied in the PUT request are true or false to either add them or remove it from the list, so I check for the parameter and its value together as a string there, so I assemble the string with the values "true" or "false". 
If needed, I can address this in a different manner

File: arangod/RestHandler/RestAdminLogHandler.cpp (hunk 16 above)
Comment (cpjulia) [reply arangod/RestHandler/RestAdminLogHandler.cpp]: Removed it

File: CHANGELOG (hunk 18 above)
Comment (cpjulia) [reply CHANGELOG]: Addressed

File: arangod/GeneralServer/RestHandler.h (hunk 19 above)
Comment (cpjulia) [reply arangod/GeneralServer/RestHandler.h]: Addressed

File: tests/js/server/aql/aql-queries-simple.js (hunk 15 above)
Comment (cpjulia) [reply tests/js/server/aql/aql-queries-simple.js]: Removed it

File: lib/Logger/LoggerFeature.h (hunk 14 above)
Comment (cpjulia) [reply lib/Logger/LoggerFeature.h:32]: Removed it

File: lib/Logger/Logger.cpp (hunk 9 above)
Comment (cpjulia) [reply lib/Logger/Logger.cpp]: Addressed

File: arangod/RestHandler/RestLogHandler.cpp (hunk 4 above)
Comment (cpjulia) [reply arangod/RestHandler/RestLogHandler.cpp]: Removed it

File: arangod/RestHandler/RestAdminLogHandler.cpp [hunk 22]
```diff
@@ -113,13 +113,27 @@ RestStatus RestAdminLogHandler::execute() {
...
+      handleLogStructuredParams();
     } else {
       generateError(rest::ResponseCode::BAD,
                     TRI_ERROR_HTTP_SUPERFLUOUS_SUFFICES,
-                    "superfluous suffix, expecting /_admin/log/entries");
+                    "superfluous suffix, expecting /_admin/log/<suffix>, "
+                    "where suffix can be either 'entries', 'level' or 'structured'");
+    }
+  } else if (type == rest::RequestType::PUT && suffixes.size() == 1) {
+    if (!suffixes.empty()) {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminLogHandler.cpp:125]: This will never be false.

File: arangod/RestHandler/RestAdminLogHandler.cpp [hunk 23]
```diff
@@ -113,13 +113,27 @@ RestStatus RestAdminLogHandler::execute() {
...
+      } else if (suffixes[0] == "structured") {
+        handleLogStructuredParams();
+      } else {
+        generateError(rest::ResponseCode::BAD,
+                      TRI_ERROR_HTTP_SUPERFLUOUS_SUFFICES,
+                      "superfluous suffix, expecting /_admin/log/<suffix>, "
+                      "where suffix can be either 'level' or 'structured'");
+      }
     }
-  } else {
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminLogHandler.cpp:121]: This case should be handled. If someone uses a wrong HTTP method or adds extra suffixes to the URL, the server won't react properly.

File: lib/Logger/LogStructuredParamsAllowList.h [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <string>
+#include <unordered_set>
+
+namespace arangodb {
+namespace structuredParams {
+// the parameters will be converted to lowercase when parsed, so the allow list
+// is in lowercase too
+const std::unordered_set<std::string> allowList = {
+    "database", "username", "url",         "collection", "queryid",
```
Comment (jsteemann) [inline lib/Logger/LogStructuredParamsAllowList.h:34]: use constexprs from below

Comment (jsteemann) [reply lib/Logger/LogStructuredParamsAllowList.h:34]: The parameters we don't support yet should be removed here.

Comment (jsteemann) [reply lib/Logger/LogStructuredParamsAllowList.h:34]: Let's try to use frozen:
``` 
constexpr auto kAllowedParameters = frozen::make_unordered_set<frozen::string>(...);
```

File: lib/Logger/Logger.cpp [hunk 25]
```diff
@@ -114,6 +116,8 @@ void LogMessage::shrink(std::size_t maxLength) {
 std::atomic<bool> Logger::_active(false);
 std::atomic<LogLevel> Logger::_level(LogLevel::INFO);
 
+std::unordered_set<std::string> Logger::_structuredLogParams({});
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:119]: Let's use a different type, e.g. std::string_view

File: lib/Logger/Logger.cpp [hunk 26]
```diff
@@ -154,6 +158,11 @@ LogGroup& Logger::defaultLogGroup() { return ::defaultLogGroupInstance; }
 
 LogLevel Logger::logLevel() { return _level.load(std::memory_order_relaxed); }
 
+std::unordered_set<std::string> Logger::structuredLogParams() {
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:161]: Can we change the set member type here as well?

File: lib/Logger/LoggerFeature.cpp [hunk 27]
```diff
@@ -416,6 +421,9 @@ void LoggerFeature::prepare() {
       std::max<uint32_t>(256, _maxEntryLength));
 
   Logger::setLogLevel(_levels);
+  if (!_structuredLogParams.empty()) {
```
Comment (jsteemann) [inline lib/Logger/LoggerFeature.cpp:424]: The if can be removed

File: tests/js/client/server_parameters/log-structured-params.js [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+
+const jsunity = require('jsunity');
+
+function LoggerSuite() {
+  'use strict';
+
+  return {
+
+    testStartingParameters: function () {
+      let res = arango.GET("/_admin/log/structured");
```
Comment (jsteemann) [inline tests/js/client/server_parameters/log-structured-params.js:45]: this test depends on order.

File: tests/js/client/server_parameters/log-structured-params.js [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+    },
+
+    testLogNewEntries: function() {
+      const res = arango.PUT("/_admin/log/structured", {"dog": true, "database": false});
+
+      assertTrue(res.hasOwnProperty("url"));
+      assertTrue(res.hasOwnProperty("username"));
+      assertFalse(res.hasOwnProperty("dog"));
+      assertFalse(res.hasOwnProperty("database"));
+    },
```
Comment (jsteemann) [inline tests/js/client/server_parameters/log-structured-params.js:103]: This test modifies the global state and doesn't clean up.

File: tests/js/client/server_parameters/log-structured-params.js [hunk 30]
```diff
@@ -0,0 +1,109 @@
+/*jshint globalstrict:false, strict:false */
```
Comment (jsteemann) [inline tests/js/client/server_parameters/log-structured-params.js:1]: We must also add a test for the JSON-based log output, i.e. when using `--log.use-json-format=true` as a startup option.
This requires an extra test file.

File: arangod/RestHandler/RestAdminLogHandler.cpp [hunk 31]
```diff
@@ -113,15 +113,43 @@ RestStatus RestAdminLogHandler::execute() {
...
+    if (suffixes.size() == 1) {
+      if (suffixes[0] == "level") {
+        handleLogLevel();
+      } else if (suffixes[0] == "structured") {
+        handleLogStructuredParams();
+      } else {
+        generateError(rest::ResponseCode::BAD,
+                      TRI_ERROR_HTTP_SUPERFLUOUS_SUFFICES,
+                      "superfluous suffix, expecting /_admin/log/<suffix>, "
+                      "where suffix can be either 'level' or 'structured'");
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminLogHandler.cpp:134]:
```suggestion
                      "where suffix can be either 'entries', 'level' or 'structured'");
```

File: lib/Logger/Logger.cpp [hunk 32]
```diff
@@ -553,21 +614,28 @@ void Logger::log(char const* logid, char const* function, char const* file,
...
-          }
-        });
-    logContext.visit(visitor);
+    {
+      READ_LOCKER(guard, _structuredParamsLock);
+      // meta data from log context
+      LogContext::OverloadVisitor visitor([&out, &dumper](
+                                              std::string_view const& key,
+                                              auto&& value) {
+        if (!_structuredLogParams.contains(key.data())) {
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:623]: Not sure if `key` is always null-terminated here. Can we also pass the size of `key` when doing the lookup?
```suggestion
        if (!_structuredLogParams.contains(key.data())) {
```

File: lib/Logger/Logger.cpp [hunk 33]
```diff
@@ -695,21 +763,27 @@ void Logger::log(char const* logid, char const* function, char const* file,
...
-          }
-          out.append("] ", 2);
-        });
-    logContext.visit(visitor);
+    {
+      READ_LOCKER(guard, _structuredParamsLock);
+      //  meta data from log
+      LogContext::OverloadVisitor visitor([&out](std::string_view const& key,
+                                                 auto&& value) {
+        if (!_structuredLogParams.contains(key.data())) {
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:771]: Same here, let's also pass the size of `key` when doing the lookup.

File: lib/Logger/Logger.cpp [hunk 34]
```diff
@@ -218,12 +228,63 @@ void Logger::setLogLevel(std::string const& levelName) {
...
+  }
+}
+
 void Logger::setLogLevel(std::vector<std::string> const& levels) {
   for (auto const& level : levels) {
     setLogLevel(level);
   }
 }
 
+std::unordered_map<std::string, bool> const Logger::parseStringParams(
```
Comment (jsteemann) [inline lib/Logger/Logger.cpp:253]:
```suggestion
std::unordered_map<std::string, bool> Logger::parseStringParams(
```

File: lib/Logger/Logger.h [hunk 35]
```diff
@@ -255,19 +259,24 @@ class Logger {
  public:
   static LogGroup& defaultLogGroup();
   static LogLevel logLevel();
+  static std::unordered_set<std::string> structuredLogParams();
   static std::vector<std::pair<std::string, LogLevel>> logLevelTopics();
   static void setLogLevel(LogLevel);
   static void setLogLevel(std::string const&);
   static void setLogLevel(std::vector<std::string> const&);
-
+  static std::unordered_map<std::string, bool> const parseStringParams(
```
Comment (jsteemann) [inline lib/Logger/Logger.h:267]:
```suggestion
  static std::unordered_map<std::string, bool> parseStringParams(
```

File: lib/Logger/LoggerFeature.cpp [hunk 36]
```diff
@@ -119,6 +119,11 @@ void LoggerFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
                   "escape unicode characters when logging",
                   new BooleanParameter(&_useUnicodeEscaped))
       .setIntroducedIn(30900);
+  options
+      ->addOption("--log.structured-param",
+                  "structured log displaying selected values",
```
Comment (jsteemann) [inline lib/Logger/LoggerFeature.cpp:124]:
```suggestion
                  "toggle usage of log category parameter in structurd log messages",
```

File: tests/js/client/server_parameters/log-structured-params-json.js [hunk 37]
```diff
@@ -0,0 +1,117 @@
...
+    },
+    testLogRestoreParams: function() {
+      const res = arango.PUT("/_admin/log/structured", {"database": true});
+
+      assertTrue(Object.keys(res).length === 3);
+      assertTrue(res.hasOwnProperty("url"));
+      assertTrue(res.hasOwnProperty("username"));
+      assertFalse(res.hasOwnProperty("dog"));
+      assertTrue(res.hasOwnProperty("database"));
+    },
```
Comment (jsteemann) [inline tests/js/client/server_parameters/log-structured-params-json.js:112]: These tests modify the global log settings and do not clean up after themselves. So the execution of tests in this file is fully order-dependent. This shouldn't be the case. If a test modifies global settings, it must clean up after itself.

In addition, these tests do not check whether the modified settings have any effect in newly created log messages.

File: tests/js/client/server_parameters/log-structured-params.js [hunk 38]
```diff
@@ -0,0 +1,117 @@
...
+      const res = arango.PUT("/_admin/log/structured", {"dog": true, "database": false});
+
+      assertTrue(res.hasOwnProperty("url"));
+      assertTrue(res.hasOwnProperty("username"));
+      assertFalse(res.hasOwnProperty("dog"));
+      assertFalse(res.hasOwnProperty("database"));
+    },
+    testLogRestoreParams: function() {
+      const res = arango.PUT("/_admin/log/structured", {"database": true});
+
```
Comment (jsteemann) [inline tests/js/client/server_parameters/log-structured-params.js:106]: Same here. These tests do not clean up after themselves so they make all tests in this file order-dependent.

File: lib/Logger/Logger.cpp (hunk 32 above)
Comment (cpjulia) [reply lib/Logger/Logger.cpp]: Addressed

File: lib/Logger/Logger.cpp (hunk 33 above)
Comment (cpjulia) [reply lib/Logger/Logger.cpp]: Addressed

File: lib/Logger/Logger.h (hunk 35 above)
Comment (cpjulia) [reply lib/Logger/Logger.h]: Addressed

File: tests/js/client/server_parameters/log-structured-params.js (hunk 38 above)
Comment (cpjulia) [reply tests/js/client/server_parameters/log-structured-params.js]: Addressed
